src/examples/librpmem/fibonacci/fibonacci.c

```diff
@@ -0,0 +1,403 @@
+	int ret;
+
+	/* map local pool */
+	size_t mapped_len;
+	void *pool = map_pmem(path, &mapped_len);
+
+	/* open remote pool */
+	RPMEMpool *rpp = remote_create_or_open(target, poolset, pool, &created);
+	if (!rpp) {
+		ret = 1;
+		goto unmap;
+	}
+
+	if (!created) {
+		/* restore from remote */
+		ret = remote_read(rpp, &fibo_r);
+		if (ret)
+			/* invalidate remote copy */
+			memset(&fibo_r, 1, DATA_SIZE);
+	} else
+		/* zero remote copy */
+		memset(&fibo_r, 0, DATA_SIZE);
+
+	/* validate copies from local and remote */
+	fibo_cleanup(rpp, pool, &fibo_r, &initialized);
+	/* generate new number */
+	if (!initialized)
+		fibo_generate(pool);
+	fibo_print(pool);
+
+	/* store to remote */
+	ret = remote_write(rpp);
+	if (ret)
+		goto err;
+
+	printf("rerun application to generate next Fibonacci number.\n");
+
+err:
+	/* close the remote pool */
+	if (rpmem_close(rpp)) {
+		fprintf(stderr, "rpmem_close: %s\n", rpmem_errormsg());
+	}
+
+unmap:
+	pmem_unmap(pool, mapped_len);
+
+	return ret;
+}
+
+static void
+parse_args(int argc, char *argv[], const char **target, const char **poolset,
+		const char **path)
+{
+	if (argc < 4) {
+		fprintf(stderr, "usage:\t%s <target> <poolset> <path>\n",
+				argv[0]);
+		exit(1);
+	}
+
+	*target = argv[1];
+	*poolset = argv[2];
+	*path = argv[3];
+}
+
+static void *
+map_pmem(const char *path, size_t *mapped_len)
+{
+	int is_pmem;
+	void *pmem = pmem_map_file(path, 0, 0, 0, mapped_len, &is_pmem);
+	if (!pmem)
+		return NULL;
+
+	if (!is_pmem) {
+		fprintf(stderr, "is not persistent memory: %s\n", path);
+		goto err;
+	}
+
+	if (*mapped_len < POOL_SIZE) {
+		fprintf(stderr, "too small: %ld < %ld\n", *mapped_len,
+				POOL_SIZE);
+		goto err;
+	}
+
+	return pmem;
+err:
+	pmem_unmap(pmem, *mapped_len);
+	exit(1);
+}
+
+/*
+ * remote_create_or_open -- create or open the remote replica
+ */
+static RPMEMpool *
+remote_create_or_open(const char *target, const char *poolset, void *pool,
+		int *created)
+{
+	struct rpmem_pool_attr pool_attr;
+	unsigned nlanes = NLANES;
+	*created = 1;
+	RPMEMpool *rpp;
+
+	/* fill pool_attributes */
+	memset(&pool_attr, 0, sizeof(pool_attr));
+	strncpy(pool_attr.signature, POOL_SIGNATURE, RPMEM_POOL_HDR_SIG_LEN);
+
+	/* create a remote pool */
+	rpp = rpmem_create(target, poolset, pool, POOL_SIZE, &nlanes,
+			&pool_attr);
+	if (rpp)
+		goto verify;
+	else {
+		if (errno != EEXIST) {
+			fprintf(stderr, "rpmem_create: %s\n",
+					rpmem_errormsg());
+			return NULL;
+		}
+		/* the remote pool exists */
+		*created = 0;
+	}
+
+	/* open a remote pool */
+	rpp = rpmem_open(target, poolset, pool, POOL_SIZE, &nlanes,
+			&pool_attr);
+	if (!rpp) {
+		fprintf(stderr, "rpmem_open: %s\n", rpmem_errormsg());
+		return NULL;
+	}
+
+verify:
+	/* verify signature */
+	if (strcmp(pool_attr.signature, POOL_SIGNATURE) != 0) {
+		pool_attr.signature[RPMEM_POOL_HDR_SIG_LEN - 1] = '\0';
+		fprintf(stderr, "invalid signature: %s != %s\n",
+				pool_attr.signature, POOL_SIGNATURE);
```

Comment:
If the signature doesn't match the expectation it's very likely this is not a pool file. This means we'll try to interpret binary data as string and possibly corrupt the user's terminal.